### ISSUE TITLE
[Flytectl Feature] Add field selector for all get commands  

### ISSUE THREAD

**Issue by pmahindrakar-oss** [flyteorg/flyte#736]
**Describe the feature/command for flytectl**
Currently the get list commands for various flyte resources including project, tasks, workflows, launchplans, executions etc
return the entire result set or are bounded through Limit  query inorder to not fetch the entire dataset for that object type.

The ask of this feature is to implement a field selector interface for flytectl which allows to narrow down the result set matching the field selector criteria.

This is similar to kubectl field selector.
https://kubernetes.io/docs/concepts/overview/working-with-objects/field-selectors/

**Provide a possible output or ux example**
Once implemented, user should be able to fetch  project, tasks, workflows,launchplans,executions  based on the field selector.
Invalid field selector should throw an error or provide details on whats fields are available for selection.

eg:
This one returns all the executions in the domain and project.
```
{
    "executions": [
        {
            "id": {
                "project": "flytesnacks",
                "domain": "development",
                "name": "er7xaa2vxf"
            },
            "spec": {
                "launch_plan": {
                    "resource_type": "LAUNCH_PLAN",
                    "project": "flytesnacks",
                    "domain": "development",
                    "name": "recipes.core.basic.lp.go_greet",
                    "version": "v103"
                },
                "metadata": {
                    "principal": "flyteconsole",
                    "system_metadata": {}
                }
            },
            "closure": {
                "outputs": {
                    "uri": "s3://my-s3-bucket/metadata/propeller/flytesnacks-development-er7xaa2vxf/end-node/data/0/outputs.pb"
                },
                "phase": "SUCCEEDED",
                "started_at": "2021-02-17T07:10:09.271487Z",
                "duration": "43.817508900s",
                "created_at": "2021-02-17T07:10:09.234425400Z",
                "updated_at": "2021-02-17T07:10:53.088995900Z",
                "workflow_id": {
                    "resource_type": "WORKFLOW",
                    "project": "flytesnacks",
                    "domain": "development",
                    "name": "recipes.core.basic.lp.go_greet",
                    "version": "v103"
                }
            }
        },
        {
            "id": {
                "project": "flytesnacks",
                "domain": "development",
                "name": "oeh94k9r2r"
            },
            "spec": {
                "launch_plan": {
                    "resource_type": "LAUNCH_PLAN",
                    "project": "flytesnacks",
                    "domain": "development",
                    "name": "recipes.core.07_juptyer_notebooks.whole_square_workflow.whole_square_wf",
                    "version": "v58"
                },
                "metadata": {
                    "principal": "flyteconsole",
                    "system_metadata": {}
                }
            },
            "closure": {
                "abort_metadata": {},
                "phase": "FAILED",
                "started_at": "2021-02-05T14:50:00.598630500Z",
                "duration": "27.059172300s",
                "created_at": "2021-02-05T14:50:00.470665800Z",
                "updated_at": "2021-02-05T14:50:27.657802300Z",
                "workflow_id": {
                    "resource_type": "WORKFLOW",
                    "project": "flytesnacks",
                    "domain": "development",
                    "name": "recipes.core.07_juptyer_notebooks.whole_square_workflow.whole_square_wf",
                    "version": "v58"
                }
            }
        }
    ]
}
```
Now with field selector i should be able to specify any field in the json and be able to filter based on it. 
for eg on the phase field. this should result in returning just one matching the phase criterial w.r.t above data
bin/flytectl get executions --field-selector=phase=SUCCEEDED

```
{
    "executions": [
        {
            "id": {
                "project": "flytesnacks",
                "domain": "development",
                "name": "er7xaa2vxf"
            },
            "spec": {
                "launch_plan": {
                    "resource_type": "LAUNCH_PLAN",
                    "project": "flytesnacks",
                    "domain": "development",
                    "name": "recipes.core.basic.lp.go_greet",
                    "version": "v103"
                },
                "metadata": {
                    "principal": "flyteconsole",
                    "system_metadata": {}
                }
            },
            "closure": {
                "outputs": {
                    "uri": "s3://my-s3-bucket/metadata/propeller/flytesnacks-development-er7xaa2vxf/end-node/data/0/outputs.pb"
                },
                "phase": "SUCCEEDED",
                "started_at": "2021-02-17T07:10:09.271487Z",
                "duration": "43.817508900s",
                "created_at": "2021-02-17T07:10:09.234425400Z",
                "updated_at": "2021-02-17T07:10:53.088995900Z",
                "workflow_id": {
                    "resource_type": "WORKFLOW",
                    "project": "flytesnacks",
                    "domain": "development",
                    "name": "recipes.core.basic.lp.go_greet",
                    "version": "v103"
                }
            }
        }
    ]
}
```

Following doc will help in finding how the filters work in admin service which flytectl communicates with for performing its operations.
https://flyte.readthedocs.io/en/latest/dive_deep/admin_service.html#using-the-admin-service

**Comment by katrogan** [comment]
hey @pmahindrakar-oss if it's helpful, here's the documentation on how how filters in the service endpoint work: https://flyte.readthedocs.io/en/latest/dive_deep/admin_service.html#using-the-admin-service

**Comment by pmahindrakar-oss** [comment]
Thanks @katrogan . Yes thats really useful. I have added the link to it in the main description aswell.

**Comment by kumare3** [comment]
cc @evalsocket if you are looking for some items to pick-up

**Comment by evalsocket** [comment]
@pmahindrakar-oss @katrogan @kumare3 is it ok if we use flyteadmin service filtering syntax like
```
flytectl get executions --field-selector="gte(duration, 100)+value_in(phase,RUNNING;SUCCEEDED;FAILED)"
```

**Comment by pmahindrakar-oss** [comment]
> @pmahindrakar-oss @katrogan @kumare3 is it ok if we use flyteadmin service filtering syntax like
> 
> ```
> flytectl get executions --field-selector="gte(duration, 100)+value_in(phase,RUNNING;SUCCEEDED;FAILED)"
> ```

According to me that sounds right. 
You can also look at the tests for more examples 
https://github.com/flyteorg/flyteadmin/blob/master/pkg/manager/impl/util/filters_test.go
https://github.com/flyteorg/flyteadmin/blob/master/pkg/common/filters_test.go

And the corr doc in the flytedocs https://docs.flyte.org/en/latest/dive_deep/admin_service.html#using-the-admin-service

**Comment by katrogan** [comment]
is the end goal of this to mimic kubectl when it comes to the field selector? If so, I don't know that passing an inline string of raw filters really fits the bill. I think we can improve on the API filter syntax as part of flytectl UX

**Comment by pmahindrakar-oss** [comment]
Having the field selector like kubectl would be an ideal way to do it as that would make anyone familiar with kubectl feel at home and much easier to play around with.

But since we currently support one type of filtering language within admin , we can choose to use it for first iteration and later enhance it to use the kubectl way OR if we prefer to directly go with new model (due to burden of backward compat etc) then we can start with making changes to admin filtering. @kumare3 @katrogan please evaluate

**Comment by brucearctor** [comment]
@pmahindrakar-oss -- I wonder the extent that flyteadmin needs to change?  Can't this just be additional functionality that gets built into flytectl?  I guess more a question for long-term design of flyteadmin?  Looking for info from others much more familiar to the design/vision.

**Comment by kumare3** [comment]
The filtering language in FlyteAdmin includes all operations and was designed by looking at many options and seems to be a common pattern

**Comment by brucearctor** [comment]
Based on @kumare3 response above  -- Sounds like FlyteAdmin not needed to change and FlyteCTL can work with as-is.

**Comment by pmahindrakar-oss** [comment]
I am seeing from kubectl users point of view.
eg  :  Following is the comparison between the two way for some relational and set operations for flyte resources.


| Operation | Kubectl  Way | FlyteAdmin  Way|
| ------------- | ------------- | ------------- |
| launch_plan.version with value 1234  | launch_plan.version == 1234 | eq(launch_plan.version, 1234)  |
| launch_plan.name with value not equal to baz  | launch_plan.name != baz | ne(launch_plan.name, baz) | 
| phase having values RUNNING, SUCCEEDED or  FAILED | phase in (RUNNING, SUCCEEDED, FAILED)  | value_in(phase,RUNNING;SUCCEEDED;FAILED) |  
| Combined conditions | launch_plan.version == 1234, launch_plan.name != baz, phase in (RUNNING, SUCCEEDED, FAILED)|value_in(phase,RUNNING;SUCCEEDED;FAILED)+ne(launch_plan.name, baz)+eq(launch_plan.version, 1234)|  

Here are my thoughts.

- Anyone familiar with data filtering sql can relate to kubectl syntax .
- Is a little more concise

On other hand.
-  The Admin way is more flexible as it allows to define more custom operator/functions on the data that the user can use  
-  This follows i guess odata form of api flitering 

I don't have a preference of one over other but would be ideal if we keep the same language across all interfaces.
i.e from flytectl, flyte-cli, plain rest API etc. If we prefer to keep it the current admin way then i think current way of passing the query in as a string  to flytectl should be fine.

**Comment by brucearctor** [comment]
I thought purpose of flytectl was to mimic what is in Kubectl.  In that light, I think you'd want the conventions the same.  Under the hood, we should be able to write code that transforms the Kubectl convention to whatever required by flyteadmin?  Or, am I missing the point?

**Comment by katrogan** [comment]
+1 @brucearctor. the Flyte admin api was designed given limitations in grpc gateway wrt repeating parameters. I'm not sure that the current syntax is necessarily the most intuitive or user friendly way to filter. If we want to make Flytectl similar to kubectl I think it's fair to expose a different syntax, especially to reduce the overhead for new users familiar with kubectl. 

Consistency within flyte is nice to have but I'd argue that programmatically accessing is an infrequent power user case and making kubectl more intuitive is a sufficient reason for departing from the admin syntax.

**Comment by pmahindrakar-oss** [comment]
Sounds good @katrogan @brucearctor . 
@evalsocket you would want to add this adapter in the filters PR.

**Comment by evalsocket** [comment]
@pmahindrakar-oss if everyone is up for kubectl syntax then I can add the changes. Another option is to ship flytectl with flyteadmin syntax and then iterate on user feedback. 

The current syntax 
```
flytectl  get projects --filters="eq(project.name, "flyteexamples")" --filters="eq(project.name, "flyteexamples")"
```

**Comment by pmahindrakar-oss** [comment]
@evalsocket Lets wait on @kumare3 to check on this.

**Comment by kumare3** [comment]
@evalsocket / @pmahindrakar-oss i like the kubectl syntax. Thank you for asking my opinion. +1 foot the kubectl syntax

**Comment by evalsocket** [comment]
@kumare3 @pmahindrakar-oss @katrogan need your feedback on user experience....New filter flag would be

- contains - `--field-selector=` ? @pmahindrakar-oss 
- gt (greater than) - `--field-selector=project.Value>4`
- gte (greter than or equal to) - `--field-selector=project.Value>=4`
- lt (less than) - `--field-selector=project.Value<5`
- lte (less than or equal to) - `--field-selector=project.Value<=5`
- eq (equal) - `--field-selector=project.name="X"`
- ne (not equal) - `--field-selector=project.name!="X"`
- value_in (for repeated sets of values) - `--field-selector=project.Phase==[RUNNING,FAILED]` ?

**Comment by brucearctor** [comment]
+1 Kubectl syntax

**Comment by pmahindrakar-oss** [comment]
Most of them seem correct except for these two and additionally we want composite queries too with comma separation which is basically logical AND 

- value_in
Currently kubectl doesn’t support set (value_in  in flyteland) operations on field selectors but we can refer to how they have done for the labels and eventually this might moved over there too

Refer : https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/
 https://github.com/kubernetes/kubernetes/issues/49387#issuecomment-782136584

kubectl get pods -n flyte -o wide --show-labels -l 'app in (minio,postgres)'
NAME                        READY   STATUS    RESTARTS   AGE   IP           NODE             NOMINATED NODE   READINESS GATES   LABELS
minio-7b985c5b58-8wdj2      1/1     Running   0          44h   10.1.15.97   docker-desktop   <none>           <none>            app=minio,pod-template-hash=7b985c5b58
postgres-6c64cd5c48-24ssd   1/1     Running   0          44h   10.1.15.94   docker-desktop   <none>           <none>            app=postgres,pod-template-hash=6c64cd5c48

- contains 
 I see contains also is not supported. contains is basically LIKE query in sql where we we dont do an exact match on the field value but just look for if the string in contained in the value.

This is implemented in admin also as 
Contains:           "field LIKE ?",
Which i believe translates or used as 
Contains:           "%value%" similar to SQL . Please check into the details by referring to the unit test i have posted earlier.
(https://github.com/flyteorg/flyteadmin/blob/master/pkg/common/filters_test.go)

@katrogan @kumare3  Please have a look.

**Comment by pmahindrakar-oss** [comment]
Also may be later we also want to support JSONPATH filtering which also would be nice and can be done on the client side .
https://kubernetes.io/docs/reference/kubectl/jsonpath/

**Comment by evalsocket** [comment]
@pmahindrakar-oss @kumare3 @katrogan I push some changes for filters. Looking for quick feedback on `values_in` and `contains` filter.

**Comment by pmahindrakar-oss** [comment]
We should probably see if we can reuse the operators from k8 api machinery which has it own parser too.
https://github.com/kubernetes/apimachinery/blob/master/pkg/selection/operator.go

**Comment by kumare3** [comment]
That's a good idea we have a dependency on api machinery

**Comment by katrogan** [comment]
nice find with the labels & annotations syntax for value in! For contains, not sure if this is a great suggestion but perhaps we could just be really verbose and do something like: 

`--field-selector=project.Phase contains "FAIL"`

**Comment by evalsocket** [comment]
@katrogan @pmahindrakar-oss @kumare3 @brucearctor made changes as you guys suggested.  Please review my....I didn't play with a large dataset so it would be great if you guys can build manually and try on your cluster and give me feedback. 

```
- contains - `--field-selector="project.Phase contains FAIL"` 
- gt (greater than) - `--field-selector=project.Value>4`
- gte (greter than or equal to) - `--field-selector=project.Value>=4`
- lt (less than) - `--field-selector=project.Value<5`
- lte (less than or equal to) - `--field-selector=project.Value<=5`
- eq (equal) - `--field-selector=project.name="X"`
- ne (not equal) - `--field-selector=project.name!="X"`
- value_in (for repeated sets of values) - `--field-selector=project.Phase==(RUNNING;FAILED)` ?
```

**Comment by pmahindrakar-oss** [comment]
I would suggest generating the data in your local flyte using flytectl. You can register serialized tasks, workflows, launchplans and also create/ terminate executions with it. You can generate a small enough set to cover simple and complex filtering queries for the flyte entities.
